fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java

```diff
@@ -183,8 +195,74 @@ private void prepare(ConnectContext connectContext, OptExpression logicOperatorT
-        if (Config.enable_experimental_mv) {
+        // TODO: add session variable
+        if (Config.enable_experimental_mv && !optimizerConfig.isRuleBased()) {
             // TODO: register materialized views
+            // register materialized views
+            registerMaterializedViews(logicOperatorTree, connectContext);
+        }
+    }
+
+    private void registerMaterializedViews(OptExpression logicOperatorTree, ConnectContext connectContext) {
+        List<Table> tables = Utils.getAllTables(logicOperatorTree);
+
+        // include nested materialized views
+        Set<MaterializedView> relatedMvs = Sets.newHashSet();
+        Utils.getRelatedMvs(tables, relatedMvs);
+
+        for (MaterializedView mv : relatedMvs) {
+            if (!mv.isActive()) {
+                continue;
+            }
+            Set<String> partitionNamesToRefresh = Sets.newHashSet();
+            // Set<String> partitionNamesToRefresh = mv.getPartitionNamesToRefreshForMv();
+            PartitionInfo partitionInfo = mv.getPartitionInfo();
+            if (partitionInfo instanceof SinglePartitionInfo) {
+                if (!partitionNamesToRefresh.isEmpty()) {
+                    continue;
+                }
+            } else if (partitionNamesToRefresh.containsAll(mv.getPartitionNames())) {
+                // if the mv is partitioned, and all partitions need refresh,
+                // then it can not be an candidate
+                continue;
+            }
+
+            // 1. build mv query logical plan
+            String mvSql = mv.getViewDefineSql();
+            ColumnRefFactory columnRefFactory = new ColumnRefFactory();
+            Pair<OptExpression, LogicalPlan> plans = Utils.getRuleOptimizedLogicalPlan(mvSql, columnRefFactory, connectContext);
+            if (plans == null) {
+                continue;
+            }
+            OptExpression mvPlan = plans.first;
+            if (!Utils.isValidMVPlan(mvPlan)) {
+                continue;
+            }
+            if (mv.getPartitionInfo() instanceof ExpressionRangePartitionInfo && !partitionNamesToRefresh.isEmpty()) {
+                Utils.updatePartialPartitionPredicate(mv, columnRefFactory, partitionNamesToRefresh, mvPlan);
+            }
+
+            List<ColumnRefOperator> outputExpressions = plans.second.getOutputColumn();
+            MaterializationContext materializationContext = new MaterializationContext(mv, mvPlan,
+                    columnRefFactory, outputExpressions, partitionNamesToRefresh);
+
+            // generate scan mv plan
+            Database db = context.getCatalog().getDb(mv.getDbId());
+            TableName tableName = new TableName(db.getFullName(), mv.getName());
+            String selectMvSql = "select * from " + tableName.toSql();
+            Pair<OptExpression, LogicalPlan> scanMvPlans =
+                    Utils.getRuleOptimizedLogicalPlan(selectMvSql, context.getColumnRefFactory(), connectContext);
```

Comment:
Can you explain why need a `scan mv plan` here?

Reply:
because rewrite will use the scanMvExpr to construct the new logical plan, which will be used in different rules